core/src/main/java/com/linecorp/armeria/internal/common/AbortedHttpResponse.java

```diff
@@ -15,12 +15,65 @@
  */
 package com.linecorp.armeria.internal.common;
 
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
 import com.linecorp.armeria.common.HttpObject;
 import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.ResponseHeaders;
 import com.linecorp.armeria.internal.common.stream.AbortedStreamMessage;
+import com.linecorp.armeria.server.HttpResponseException;
 
 public final class AbortedHttpResponse extends AbortedStreamMessage<HttpObject> implements HttpResponse {
     public AbortedHttpResponse(Throwable cause) {
         super(cause);
     }
+
+    @Override
+    public HttpResponse mapHeaders(Function<? super ResponseHeaders, ? extends ResponseHeaders> function) {
+        requireNonNull(function, "function");
+        return mapIncludingHttpResponseException(it -> it.mapHeaders(function));
+    }
+
+    @Override
+    public HttpResponse mapTrailers(
+            Function<? super HttpHeaders, ? extends HttpHeaders> function) {
+        requireNonNull(function, "function");
+        return mapIncludingHttpResponseException(it -> it.mapTrailers(function));
+    }
+
+    @Override
+    public HttpResponse mapData(
+            Function<? super HttpData, ? extends HttpData> function) {
+        requireNonNull(function, "function");
+        return mapIncludingHttpResponseException(it -> it.mapData(function));
+    }
+
+    @Override
+    public HttpResponse mapInformational(
+            Function<? super ResponseHeaders, ? extends ResponseHeaders> function) {
+        requireNonNull(function, "function");
+        return mapIncludingHttpResponseException(it -> it.mapInformational(function));
+    }
+
+    private HttpResponse mapIncludingHttpResponseException(Function<HttpResponse, HttpResponse> function) {
+        requireNonNull(function, "function");
+        if (getCause() instanceof HttpResponseException) {
+            final HttpResponseException httpResponseException = (HttpResponseException) getCause();
+            if (httpResponseException.getCause() == null) {
+                return function.apply(httpResponseException.httpResponse());
+            } else {
+                return new AbortedHttpResponse(
+                        HttpResponseException.of(
+                                function.apply(httpResponseException.httpResponse()),
+                                httpResponseException.getCause()
+                        )
+                );
+            }
+        }
+        return function.apply(this);
```

Comment:
How about just returning `this` because `function` is useless if the cause is not an instance of `HttpResponseException`?
I think we can even raise an exception if `mapX` is called when the cause is not an instance of `HttpResponseException`.